### PR TITLE
(PA-4195) Adds Snyk scan GitHub workflow

### PR DESCRIPTION
**Pull request by mhashizume**
Testing out work from https://github.com/puppetlabs/pxp-agent-vanagon/pull/41

### CODE REVIEW

**Comment by mhashizume** [comment]
Hi @LivingInSyn , I think this repo needs a few more secrets added to it (`SEC_PUBLIC_RPROXY_KEY`, ` SEC_PUBLIC_RPROXY_USER `, and `SEC_SECBOT_SSH_PUBLIC`) in order to resolve the issues I'm seeing in the actions: https://github.com/puppetlabs/pxp-agent-vanagon/runs/5074416219?check_suite_focus=true#step:4:37

Could you verify that this is the issue? Thanks!

File: .github/workflows/snyk_scan.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      uses: puppetlabs/security-snyk-vanagon-action@v2.2.0
+      with:
+        snykToken: ${{ secrets.SNYK_PE_TOKEN }}
+        snykOrg: 'puppet-enterprise'
+        skipPlatforms: 'debian-10-armhf'
+        urlsToReplace: 'artifactory.delivery.puppetlabs.net,%s/xart,builds.delivery.puppetlabs.net,%s/xbuild'
+        newHost: 'localhost'
+        rproxyUser: ${{ secrets.RPROXY_USER }}
+        rproxyKey: ${{ secrets.RPROXY_KEY }}
+        sshKey: ${{ secrets.SEC_SECBOT_SSH_PUBLIC }}
```
Comment (LivingInSyn) [inline .github/workflows/snyk_scan.yaml:25]: `SEC_PUBLIC_RPROXY_USER` is now available
`SEC_PUBLIC_RPROXY_KEY` is now available
`SEC_SECBOT_SSH_PUBLIC` is now available

Comment (joshcooper) [reply .github/workflows/snyk_scan.yaml:25]: Thanks @LivingInSyn Unfortunately we're seeing a warning:

https://github.com/puppetlabs/pxp-agent-vanagon/runs/5101556125?check_suite_focus=true#step:4:12

```
/usr/local/bin/start_script.sh: line 21: warning: command substitution: ignored null byte in input
```

And then a fatal error trying to retrieve data from remapped build server URL:

https://github.com/puppetlabs/pxp-agent-vanagon/runs/5101556125?check_suite_focus=true#step:4:40

```
2022/02/07 23:26:29 Got bad output from vanagon dependencies on pxp-agent redhatfips-7-x86_64. Try #0 Output: Error loading project 'pxp-agent' using '/github/workspace/configs/projects/pxp-agent.rb':
Error: 401. Unable to get source from http://localhost/xbuild/puppet-runtime/202202020/artifacts/agent-runtime-main-202202020.redhatfips-7-x86_64.settings.yaml.sha1
```

It looks like the project is trying to inherit settings from the `pxp-agent` component in https://github.com/puppetlabs/pxp-agent-vanagon/blob/328192431f759bbef1efe2e8498ffff84e76d436/configs/projects/pxp-agent.rb#L27

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml:25]: Going to work on debugging this today with @jelinwils

Comment (joshcooper) [reply .github/workflows/snyk_scan.yaml:25]: The `vanagon dependencies pxp-agent <platform>` command needs to retrieve settings from the already built and promoted `puppet-runtime` as specified in:

```json
{"location":"http://builds.delivery.puppetlabs.net/puppet-runtime/202202020/artifacts/","version":"202202020"}
```

Presumably we don't have access to this in the context of a github action?

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml:25]: you do, that's what the reverse proxy is there for

Comment (joshcooper) [reply .github/workflows/snyk_scan.yaml:25]: @LivingInSyn there's a reverse proxy for `artifactory`, but not `builds`, right?

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml:25]: The reverse proxy can access both. I'll file a bug for that failure condition that reports as a success and we'll get it fixed along with the `.snyk` support

Comment (joshcooper) [reply .github/workflows/snyk_scan.yaml:25]: The reverse proxy can't resolve DNS for `builds`: https://github.com/puppetlabs/pxp-agent-vanagon/runs/5128148360?check_suite_focus=true#step:4:41

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml:25]: you're missing this from your action:

https://github.com/puppetlabs/security-snyk-vanagon-action/blob/main/sample_workflow.yaml#L25

Without `urlsToReplace` the builds url is never replaced with the reverse proxy host

Comment (mhashizume) [reply .github/workflows/snyk_scan.yaml:25]: This branch has the `urlsToReplace` and `newHost` keys, but running the action still fails: https://github.com/puppetlabs/pxp-agent-vanagon/runs/5101556125?check_suite_focus=true#step:4:41

Those two values are the same as the sample workflow in the action repo. Should those be different? I'm guessing that I need the reverse proxy URL in the `newHost` key (which is currently set to `localhost`)?

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml]: @mhashizume the error you just referenced is a 401 coming from the reverse proxy:

https://github.com/puppetlabs/pxp-agent-vanagon/runs/5101556125?check_suite_focus=true#step:4:41

you are referencing the wrong secrets here:

https://github.com/puppetlabs/pxp-agent-vanagon/actions/runs/1797792732/workflow#L23-L24

The ones added to this repo are:

`SEC_PUBLIC_RPROXY_KEY` and `SEC_PUBLIC_RPROXY_USER`

Comment (mhashizume) [reply .github/workflows/snyk_scan.yaml]: Thanks for catching that. I updated those variables in the workflow a little earlier today and seem to still be encountering the same issue from the reverse proxy.

You can see the correct variable names in this workflow: https://github.com/puppetlabs/pxp-agent-vanagon/actions/runs/1820101684/workflow

Still result in a 401 error: https://github.com/puppetlabs/pxp-agent-vanagon/runs/5130999380?check_suite_focus=true#step:4:42

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml]: That's actually a `Error: 404` now, which is good, it's a _different_ error 😂  Lets setup some time tomorrow to just jam this out and get y'all moving

Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml]: actually, it successfully completed the following project+platforms
```
2022/02/09 19:48:01 running snyk on pxp-agent ubuntu-20.04-amd64
2022/02/09 19:48:01 testing  pxp-agent ubuntu-20.04-amd64
2022/02/09 19:48:01 running snyk on pxp-agent ubuntu-20.04-aarch64
2022/02/09 19:48:01 testing  pxp-agent ubuntu-20.04-aarch64
2022/02/09 19:48:01 running snyk on pxp-agent sles-15-x86_64
2022/02/09 19:48:01 testing  pxp-agent sles-15-x86_64
2022/02/09 19:48:01 running snyk on pxp-agent ubuntu-16.04-i386
2022/02/09 19:48:01 running snyk on pxp-agent solaris-11-sparc
2022/02/09 19:48:01 running snyk on pxp-agent solaris-11-i386
2022/02/09 19:48:01 testing  pxp-agent solaris-11-i386
2022/02/09 19:48:01 testing  pxp-agent ubuntu-16.04-i386
2022/02/09 19:48:01 testing  pxp-agent solaris-11-sparc
2022/02/09 19:48:06 finished snyk testing:  pxp-agent sles-15-x86_64
2022/02/09 19:48:06 Finished running snyk on pxp-agent sles-15-x86_64
2022/02/09 19:48:07 finished snyk testing:  pxp-agent ubuntu-20.04-aarch64
2022/02/09 19:48:07 Finished running snyk on pxp-agent ubuntu-20.04-aarch64
2022/02/09 19:48:08 finished snyk testing:  pxp-agent ubuntu-16.04-i386
2022/02/09 19:48:08 Finished running snyk on pxp-agent ubuntu-16.04-i386
2022/02/09 19:48:08 finished snyk testing:  pxp-agent ubuntu-20.04-amd64
2022/02/09 19:48:08 Finished running snyk on pxp-agent ubuntu-20.04-amd64
2022/02/09 19:48:08 finished snyk testing:  pxp-agent solaris-11-i386
2022/02/09 19:48:08 Finished running snyk on pxp-agent solaris-11-i386
2022/02/09 19:48:08 finished snyk testing:  pxp-agent solaris-11-sparc
2022/02/09 19:48:08 Finished running snyk on pxp-agent solaris-11-sparc
```
The others are 404's from artifactory or builds

File: .github/workflows/snyk_scan.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+---
+name: Vanagon Snyk scan
+
+on: [pull_request]
```
Comment (LivingInSyn) [inline .github/workflows/snyk_scan.yaml:4]: This should just be `on: merge` now given our recent conversations around `pull_request` and `pull_request_target`

Comment (joshcooper) [reply .github/workflows/snyk_scan.yaml:4]: did you mean `on: push`?

**Comment by joshcooper** [comment]
Are these failures expected...

```2022/02/15 21:59:39 Got bad output from vanagon dependencies on pxp-agent ubuntu-20.04-amd64. Try #1 Output: An error was encountered evaluating block. Retrying..
```

File: .github/workflows/snyk_scan.yaml (hunk 2 above)
Comment (LivingInSyn) [reply .github/workflows/snyk_scan.yaml:4]: 🤦 yes, sorry

**Review by joshcooper — APPROVED** [review]

**Comment by LivingInSyn** [comment]
@joshcooper yes, some failures are expected. It'll try 3 times to run `vanagon dependencies` on each project+platform. Failures that say `total failure on: ` should be investigated or raised to us.

**Comment by joshcooper** [comment]
@LivingInSyn There is a total failure for [`el-8-x86_64`]( https://github.com/puppetlabs/pxp-agent-vanagon/runs/5220839125?check_suite_focus=true#step:4:84) and there is no entry in snyk.io for that platform, yet the action passed. Seems like a bug in the action.

If I run `bundle exec vanagon dependencies pxp-agent el-8-x86_64` it runs successfully. However, `vanagon` runs the following command to see if a repo exists:

```
Attempting to fetch from canonical URL "https://github.com/puppetlabs/pxp-agent.git"
Executing 'git ls-remote https://github.com/puppetlabs/pxp-agent.git > /dev/null 2>&1' locally
```

This is an expensive way to check as github will return the HEAD ref for all branches and tags:

```
$ git ls-remote https://github.com/puppetlabs/pxp-agent.git | wc -l
1087
```

I think github is rate-limiting us, leading to transients, so I'll file a VANAGON ticket. However, it would help if the action provided more information about what commands were being run and what failed, as well as failing the run if there are any total failures.

**Comment by joshcooper** [comment]
We're going to look at solving this a different way, closing for now.

**Pull request closed without merging**